### PR TITLE
Handle closure compiler type cast syntax correctly

### PR DESCRIPTION
**Pull request by yangsu**
Fixes https://github.com/prettier/prettier/issues/1445

### CODE REVIEW

File: src/comments.js [hunk 1]
```diff
@@ -990,6 +995,12 @@ function printComments(path, print, options, needsSemi) {
       const text = options.originalText;
       if (util.hasNewline(text, util.skipNewline(text, util.locEnd(comment)))) {
         leadingParts.push(hardline);
+      } else if (util.isClosureCompilerTypeCastComment(text, comment)) {
```
Comment (azz) [inline src/comments.js]: I'm wondering if this can instead be handled in `fast-path.js`'s `needsParens` function.

**Comment by azz** [comment]
Any idea how this will play with [flow comment types](https://flow.org/en/docs/types/comments/)?

File: src/comments.js (hunk 1 above)
Comment (vjeux) [inline src/comments.js]: Agreed. Could you look into it? If that works that would be cleaner.

File: src/util.js [hunk 2]
```diff
@@ -445,6 +445,16 @@ function isBlockComment(comment) {
   return comment.type === "Block" || comment.type === "CommentBlock";
 }
 
+function isClosureCompilerTypeCastComment(text, comment) {
+  // https://github.com/google/closure-compiler/wiki/Annotating-Types#type-casts
+  // Syntax example: var x = /** @type {string} */ (fruit);
+  return (
+    isBlockComment(comment) &&
+    comment.value.match(/^\*\s+@type\s+{[^}]+}\s+$/) &&
```
Comment (vjeux) [inline src/util.js]: probably want to use `\s*` instead of `\s+` and add some inside of `{}`

**Comment by vjeux** [comment]
Thanks for working on it!

File: src/util.js (hunk 2 above)
Comment (yangsu) [inline src/util.js]: Oops. that's what i meant to use.

**Comment by yangsu** [comment]
@azz I tested a few variations of the syntax and it seems that the changes I made to how comments are attached behave correctly.

File: src/util.js [hunk 3]
```diff
@@ -445,6 +445,20 @@ function isBlockComment(comment) {
   return comment.type === "Block" || comment.type === "CommentBlock";
 }
 
+function hasClosureCompilerTypeCastComment(text, node) {
+  // https://github.com/google/closure-compiler/wiki/Annotating-Types#type-casts
+  // Syntax example: var x = /** @type {string} */ (fruit);
+  return (
+    node.comments &&
+    node.comments.every(
```
Comment (azz) [inline src/util.js]: Should this be `every` or `some`?

**Comment by vjeux** [comment]
@azz didn't you disable codecov comments?

**Comment by azz** [comment]
@vjeux I certainly did... Maybe needs rebase?

**Comment by yangsu** [comment]
Thank you for your comments @azz @vjeux. Any other concerns?

**Comment by yangsu** [comment]
Ping @azz @vjeux

**Comment by vjeux** [comment]
I'm fine with it, I'll let @azz merge it as he did the initial review

**Comment by azz** [comment]
Heh, your comment appeared right as I clicked the merge button.

**Pull request merged**